segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -196,86 +328,302 @@ public boolean supportsTruncation() {
+                return;
+            }
+
+            // We need to be very careful with IOUtils.copyBytes. There are many overloads with very similar signatures.
+            // There is a difference between (InputStream, OutputStream, int, boolean) and (InputStream, OutputStream, long, boolean),
+            // in that the one with "int" uses the third arg as a buffer size, and the one with "long" uses it as the number
+            // of bytes to copy.
+            IOUtils.copyBytes(data, stream, (long) length, false);
+
+            stream.flush();
+        } catch (IOException ex) {
+            throw HDFSExceptionHelpers.throwException(handle.getSegmentName(), ex);
+        }
+
+        HDFSMetrics.WRITE_LATENCY.reportSuccessEvent(timer.getElapsed());
+        HDFSMetrics.WRITE_BYTES.add(length);
+        LoggerHelpers.traceLeave(log, "write", traceId, handle, offset, length);
     }
 
     @Override
-    public SegmentProperties getStreamSegmentInfo(String streamSegmentName) throws StreamSegmentException {
-        return call(new GetInfoOperation(streamSegmentName, this.context));
+    public SegmentHandle openWrite(String streamSegmentName) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        int fencedCount = 0;
+        do {
+            try {
+                FileStatus existingFiles = findStatusForSegment(streamSegmentName, true);
+
+                if (!isSealed(existingFiles.getPath())) {
+                    if (getEpocFromPath(existingFiles.getPath()) > this.epoch) {
+                        throw new StorageNotPrimaryException(streamSegmentName);
+                    }
+
+                    Path targetPath = getFilePath(streamSegmentName, this.epoch);
+                    if (!targetPath.equals(existingFiles.getPath())) {
+                        try {
+                            this.fileSystem.rename(existingFiles.getPath(), targetPath);
+                        } catch (FileNotFoundException e) {
+                            //This happens when more than one host is trying to fence and only one of the host goes through
+                            // retry the rename so that host with the highest epoch gets access.
+                            log.warn("Race in fencing. More than two hosts trying to own the segment. Retrying");
+                            fencedCount++;
+                            continue;
+                        }
+                    }
+                }
+                //Ensure that file exists
+                findStatusForSegment(streamSegmentName, true);
+                return HDFSSegmentHandle.write(streamSegmentName);
+            } catch (IOException e) {
+                throw HDFSExceptionHelpers.throwException(streamSegmentName, e);
+            }
+        } while (fencedCount < MAX_ATTEMPT_COUNT);
+        LoggerHelpers.traceLeave(log, "openWrite", traceId, epoch);
```

Comment:
What happens if fencedCount >= MAX_ATTEMPT_COUNT? Returning null is not part of the contract for this method.